src/utils/logger.ts

```diff
@@ -0,0 +1,68 @@
+import { AddressOrPair } from "@polkadot/api/submittable/types";
+import { SubmittableExtrinsic, ApiTypes } from "@polkadot/api/types";
+import { Callback, ISubmittableResult } from "@polkadot/types/types";
+import { EventRecord, DispatchError } from "@polkadot/types/interfaces/system";
+import { ApiPromise } from "@polkadot/api";
+
+declare type AnyFunctionReturningSubmittableExtrinsicApis =
+    SubmittableExtrinsic<ApiTypes>;
+
+interface EventContainer {
```

Comment:
Are we planning on adding anything to this interface? Otherwise it does feel like passing `events` directly might be enough.

Reply:
In order to return `events` directly, we'd need to wait for the callback function to be called, which isn't the most JavaScript-y way of doing things. I argue that setting the field of the passed EventContainer object is cleaner

Reply:
I think the opposite, waiting for callbacks is a very JavaScript-y way to do things.
With the current implementation, we have no way to tell when `events` is set.
I would rather suggest returning a `Promise<EventRecord[]>`.
The `signAndSend` looks like it was designed for rxjs, not for promises, so the API is really nasty but I think It should be doable with something like this

```typescript
const { unsubscribe, result } = await new Promise((resolve) => {
        let unsubscribe: () => void;
        // When passing { nonce: -1 } to signAndSend the API will use system.accountNextIndex to determine the nonce
        return transaction
            .signAndSend(signer, { nonce: -1 }, (result: ISubmittableResult) => resolve({ unsubscribe, result }))
            .then((u: (() => void)) => (unsubscribe = u));
    });
```

As with your solution, we are relying on the fact that the internal implementation should resolve the promise before calling the callback, otherwise both solutions fail.